Add perl module Text::Fuzzy

lskatz opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Gitter channel][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

lskatz committed:
perl-text-fuzzy

lskatz commented:
Note: this was originally submitted to bioconda but is probably more appropriate for conda-forge <https://github.com/bioconda/bioconda-recipes/pull/46913>

lskatz committed:
misc

lskatz committed:
add recipe-maintainers, corrected URL, and perl license

lskatz committed:
try to specify linux only

lskatz committed:
try to specify linux only

lskatz committed:
m try to specify linux only

lskatz committed:
trying to force linux still

lskatz committed:
remove skip:true

lskatz committed:
remove libcrypt-dev linux specific req

lskatz committed:
conda_build_config.yaml

lskatz committed:
remove noarch:generic

lskatz commented:
I am not clear on how to limit the build to just Linux

lskatz commented:
> I am not clear on how to limit the build to just Linux

@conda-forge/help-perl I am almost ready for review but I have no idea how to just limit the test to Linux. Thank you for any help. Additionally, when I build this locally with `conda-build .`, it fails with 

```text
lib/5.26.2/x86_64-linux-thread-multi/CORE/reentr.h:104:17: fatal error: crypt.h: No such file or directory
  104 | #       include <crypt.h>
      |                 ^~~~~~~~~
compilation terminated.
make: *** [Makefile:338: Fuzzy.o] Error 1
```

I am not sure if this is an actual issue since it seems to build in the CI environment.

lskatz commented:
I see some guidance in #25110 but I don't think it completely addresses my issues.

jvolkening commented:
Hi Lee,

To limit the build to just Linux, you need to add `skip: true  # [not linux]` under `build:` in your meta.yaml. My personal opinion is that it is preferred to try to get the OSX and Windows builds to work, but admittedly this can be a challenge with the given state of Perl in conda or without access to those platforms for testing. A few other comments on your recipe:

* The module is compiled so definitely not `noarch` -- you can remove that commented line completely.

* The current policy in conda-forge is to use `INSTALLDIRS=vendor` in your build script rather than `site`. The full recommended argument string to `Makefile.PL` is `INSTALLDIRS=vendor NO_PERLLOCAL=1 NO_PACKLIST=1`.

* All of the `host` and `run` requirements except for `perl` are unnecessary and should be removed. I believe many of the `build` requirements are also unnecessary -- see below.

* The SPDX license string for most Perl packages that indicate "licensed under the same terms as Perl" or `LICENSE => 'perl'` (including Text::Fuzzy) should be `Artistic-1.0-Perl OR GPL-1.0-or-later`.

* Conda-forge wants `license_file` specified under `about:`, so this file either needs to be packaged with your recipe or included in some other way. In my migrated recipes I've been packaging a LICENSE file when missing, but I got some push-back on this in one review, although I don't agree with the alternative suggested. For now I would suggest downloading the text of these licenses and including them in your recipe.

I've developed an alternative Perl recipe generator as part of the bioconda->conda-forge migration. In case it helps for comparison of requirements, etc, here is the recipe it generated for Text::Fuzzy which builds successfully under Linux and Windows (without `license_file` as it's still being worked out how best to handle that):

```yaml
{% set version = "0.29" %}
{% set sha256 = "3df5cfd2ca1a4c5ca7ff7bab3cc8d53ad2064e134cbf11004f3cf8c4b9055bff" %}

# regex to use for Windows build hack
{% set win_patch = "s|C:\\\\strawberry\\\\c|$ENV{LIBRARY_PREFIX}\\\\mingw-w64|g" %}

package:
  name: perl-text-fuzzy
  version: {{ version }}

source:
  url: https://cpan.metacpan.org/authors/id/B/BK/BKB/Text-Fuzzy-{{ version }}.tar.gz
  sha256: {{ sha256 }}

build:
  number: 0
  # needed for linker to find some m2w64 libs
  merge_build_host: True  # [win]
  script:
    - >-
      perl Makefile.PL INSTALLDIRS=vendor NO_PERLLOCAL=1 NO_PACKLIST=1 MAKE=make
      && (for /r . %%f in (*Makefile) do perl -i -pe "{{ win_patch }}" %%f)  # [win]
      && make
      && make test
      && make install VERBINST=1

requirements:
  build:
    - {{ compiler('c') }}  # [unix]
    - m2w64-gcc # [win]
    - make
  host:
    - perl >= 5.32
  run:
    - perl >= 5.32

test:
  imports:
    - Text::Fuzzy
  source_files:
   - t
  commands:
   - prove

about:
  home: https://metacpan.org/pod/Text::Fuzzy
  summary: Partial string matching using edit distances
  license: Artistic-1.0-Perl OR GPL-1.0-or-later

extra:
  recipe-maintainers:
    - jvolkening
    - conda-forge/perl-packagers
```

lskatz commented:
Oh wow there is so much here that I didn't know. Thank you so much for your help.

Follow up questions:

* Do I remove build.sh because now it is running Makefile.PL inside of the yaml?
* Do I remove conda_build_config.yaml because now it works in multiple OSs?
* Do you think I can reduce the perl version to something smaller like 5.16 to help with compatibility? Or I guess I can at least test it in the CI.
* Should I put my github username in recipe-maintainers?
* Do I literally copy and paste a license into license.txt and then point to it in `license_file`? Do I need a copy step or anything?

It seems to have built locally on my Linux computer and so I'm going to try it on CI!

lskatz committed:
used all suggestions from jvolkening

lskatz committed:
updated win selector; deleted last empty line; added LICENSE

lskatz committed:
Merge branch 'main' into main

lskatz commented:
@jvolkening it seems to pass CI! How do I follow through with your change to make yourself a maintainer? It says you have to comment? And then how do I trigger a merge?

jvolkening commented:
> * Do I remove build.sh because now it is running Makefile.PL inside of the yaml?

Yes, looks like you took care of this.

> * Do I remove conda_build_config.yaml because now it works in multiple OSs?

I've never made use of that file -- you shouldn't need it here I don't think.

> * Do you think I can reduce the perl version to something smaller like 5.16 to help with compatibility? Or I guess I can at least test it in the CI.

You can leave the version unspecified -- that's normally what I do but here for some reason the build test was pulling in an earlier Perl today which didn't work. In the 5.32 conda release the Windows build switched to using Strawberry Perl, and so the recipes are different to accommodate this -- the recipe I shared wouldn't build on earlier Windows conda releases. Actually, I have a PR in to the Perl feedstock to switch to a native Windows build, so if that ever gets merged things should be simpler.

> * Should I put my github username in recipe-maintainers?

Yes, I would swap mine with yours. You can leave mine in if you want, but I hope to be added to perl-packagers and then it won't matter.

> * Do I literally copy and paste a license into license.txt and then point to it in `license_file`? Do I need a copy step or anything?

How you did it looks good.

jvolkening commented on recipes/perl-text-fuzzy/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+  host:
+    - perl >=5.16
+  run:
+    - perl >=5.16
```
```
 host:
    - perl >=5.16
  run:
    - perl >=5.16
```

For the Windows build at least, this needs to be 5.32 or else remove the version altogether (preferred). The build should get pinned to the latest Perl build in any case.

lskatz replied:
I removed the version

jvolkening commented on recipes/perl-text-fuzzy/LICENSE:
```diff
@@ -0,0 +1,127 @@
+			 The "Artistic License"
```
Since it is an either/or license, to be complete you may want to include the GPL 1.0 text too. The exact requirements here are a bit unclear to me though.

lskatz replied:
Since it's unclear and I might never know what the actual requirement is until someone merges, I'll leave it as-is with a single license file.

jvolkening commented:
> it seems to pass CI! How do I follow through with your change to make yourself a maintainer? It says you have to comment? And then how do I trigger a merge?

I am okay with being included as a maintainer on this recipe. You will need to get someone from perl-packagers to comment as well. You will need someone with merge permissions for the final review and merge. Be prepared for a slow process here -- I have some outstanding PRs from back in January and February that are still waiting for final reviews. There just aren't that many people who are conda-forge admins and are willing to spend time on the Perl-related stuff.

I just did a review of the current commit, but I don't have the ability to merge.

"mfansler" and "xileF1337" have been helpful on my recipes recently. "mbargull" has done a fair amount of work on Perl recipes in the past few years. I would suggesting pinging them along with "perl-packagers" to ask for a review.

lskatz committed:
swap maintainers for myself; remove perl version

lskatz commented on recipes/perl-text-fuzzy/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+  license_file: LICENSE
+
+extra:
+  recipe-maintainers:
```
I swapped you for me, and I kept `conda-forge/perl-packagers`.

lskatz committed:
remove conda-forge/perl-packagers until they merge, to pass linting

lskatz commented:
> > it seems to pass CI! How do I follow through with your change to make yourself a maintainer? It says you have to comment? And then how do I trigger a merge?
> 
> I am okay with being included as a maintainer on this recipe. You will need to get someone from perl-packagers to comment as well. You will need someone with merge permissions for the final review and merge. Be prepared for a slow process here -- I have some outstanding PRs from back in January and February that are still waiting for final reviews. There just aren't that many people who are conda-forge admins and are willing to spend time on the Perl-related stuff.
> 
> I just did a review of the current commit, but I don't have the ability to merge.
> 
> "mfansler" and "xileF1337" have been helpful on my recipes recently. "mbargull" has done a fair amount of work on Perl recipes in the past few years. I would suggesting pinging them along with "perl-packagers" to ask for a review.

Thank you for all your help! It looks like it is passing linting. I didn't make any other major changes, and so I expect the rest of the tests to continue passing. I'll see what I can do about putting this recipe on my private channel and also how I might even make that private channel while I wait.

lskatz commented:
> "mfansler" and "xileF1337" have been helpful on my recipes recently. "mbargull" has done a fair amount of work on Perl recipes in the past few years. I would suggesting pinging them along with "perl-packagers" to ask for a review.

@mfansler @xileF1337 @mbargull @conda-forge/help-perl I would like to merge this pull request. All checks have passed. Thank you for any assistance!

lskatz commented:
I have learned how to make my own channel and so I uploaded to `-c lskatz`, e.g., `conda install -c lskatz perl-text-fuzzy` in the meantime.

lskatz committed:
Merge branch 'main' of github.com:conda-forge/staged-recipes into main

lskatz commented:
Ping

traversaro commented:
@lskatz Sorry for the long wait for a review, can you please rebase on top of the latest main to ensure that the CI is still working? Thanks!

lskatz committed:
Merge branch 'conda-forge:main' into main

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/perl-text-fuzzy/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/perl-text-fuzzy/meta.yaml**:

 * ❌ This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.

For **recipes/perl-text-fuzzy/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/19364848945](https://github.com/conda-forge/conda-forge-webservices/actions/runs/19364848945). Examine the logs at this URL for more detail.</sub>

lskatz committed:
Merge branch 'main' of github.com:lskatz/staged-recipes

lskatz committed:
compilers

lskatz committed:
help from chatgpt

lskatz committed:
removed m2w64-cc so that I can just depend on compiler('c'); add perl to build env

lskatz committed:
update win compiler to gcc

lskatz committed:
remove unnecessary tests

lskatz committed:
updated test command

lskatz committed:
remove perl from build to help it run/host from the correct path

lskatz committed:
remove strawberry perl patch

lskatz committed:
gxx compiler for windows

lskatz committed:
using a win safe method of testing perl

lskatz committed:
Merge branch 'main' into main

jvolkening commented on recipes/perl-text-fuzzy/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+{% set sha256 = "3df5cfd2ca1a4c5ca7ff7bab3cc8d53ad2064e134cbf11004f3cf8c4b9055bff" %}
+
+# regex to use for Windows build hack
+#{% set win_patch = "s|C:\\\\strawberry\\\\c|$ENV{LIBRARY_PREFIX}\\\\mingw-w64|g" %}
```
You can remove this since you removed the line where the variable is used.

lskatz replied:
Got it

jvolkening commented on recipes/perl-text-fuzzy/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+
+test:
+  commands:
+    - perl -we "use Text::Fuzzy;"
```
Ideally it would be better to run the regular test suite, but this simplified testing has been used frequently in the past and I think is okay for now.

jvolkening reviewed (changes requested):
Overall, looks good to me. Once you remove the unused `win_patch` lines, it should be ready.

lskatz committed:
Remove Windows build hack comment from meta.yaml

Removed commented-out regex for Windows build hack.

jvolkening reviewed (approved):
Looks good

lskatz committed:
try to add tests

danielnachun commented on recipes/perl-text-fuzzy/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+    - libcxx               # [unix]
+    - gcc                  # [win]
+    - gxx                  # [win]
+    - make
```
```suggestion
  build:
    - {{ compiler('c') }}
    - {{ stdlib("c") }}
    - make
```
Let's see if MSVC works for Windows here.  If that doesn't work we can try `m2w64` instead.

Also the upstream repository at https://github.com/benkasminbullock/Text-Fuzzy contains no embedded C++ code, only C.  So I'm not sure why `libcxx` was added to the dependencies but it should not be needed.

If the changes I suggested fail, please let me know before reverting anything!  I've built dozens of native Perl packages in my personal repository but we're still figuring out how to do native Perl packages on Windows right so it will be helpful to see any failures if they occur.

lskatz committed:
Update recipes/perl-text-fuzzy/meta.yaml

Co-authored-by: Daniel Nachun <daniel.nachun@gmail.com>

lskatz committed:
Update recipes/perl-text-fuzzy/meta.yaml

Let's see if MSVC works for Windows here. If that doesn't work we can try m2w64 instead.

Co-authored-by: Daniel Nachun <daniel.nachun@gmail.com>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/perl-text-fuzzy/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/perl-text-fuzzy/meta.yaml**:

 * ❌ The recipe is not parsable by any of the known recipe parsers (['conda-forge-tick (the bot)', 'conda-recipe-manager', 'conda-souschef (grayskull)']). Please check the logs for more information and ensure your recipe can be parsed.
 * ❌ package.version has a value that is interpreted as a floating-point number. Please quote it (like "0.29" or "{{ var }}") to ensure that it is interpreted as string and preserved exactly.

For **recipes/perl-text-fuzzy/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-forge-tick (the bot)`. Your recipe  may not receive automatic updates and/or may not be compatible with conda-forge's infrastructure. Please check the logs for more information and ensure your recipe can be parsed.
 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.
 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/25964183831](https://github.com/conda-forge/conda-forge-webservices/actions/runs/25964183831). Examine the logs at this URL for more detail.</sub>

lskatz committed:
Merge branch 'main' into main

lskatz commented:
Hi, pinging you all in case it's helpful. I am refreshing this branch and therefore the CI results right now.

traversaro commented:
> Hi, pinging you all in case it's helpful. I am refreshing this branch and therefore the CI results right now.

Sorry for the late reply, did you checked the comment of @danielnachun in https://github.com/conda-forge/staged-recipes/pull/25932/changes#r2670767030 ? They seem to be related to the CI failures.
